### PR TITLE
Feature/af 195 add helm orchestrator

### PR DESCRIPTION
**Pull request by benjamin-wiffen**

### CODE REVIEW

**Review by cameron-thomson-bsl — APPROVED** [review]

File: README.md [hunk 1]
```diff
@@ -139,6 +139,96 @@ command groups. The `NullOrchestrator` disables commands related to managing ser
...
+    # [Optional] Default is `chart/`
+    chart_location="mychart.tgz",
+
+    # Chart source path (absolute).
+    # [Optional] Default is `None`.
+    dev_chart_location="path/to/mychart"
+)
+```
+
+_NOTE:_ Using `dev_chart_location` allows you to specify the location to your charts source code, so that live changes can be reflected every time `./myapp service start` is run.
```
Comment (al-jeyapal) [inline README.md:184]: This needs to be clearer. Explain when/why you would use it.
Presumably it is ONLY for development purposes, the final version should never include it.
In which case, we're better off not having it as a config value here, and rather using a ENV var to handle it.

```
MYAPP_DEV_MODE=true MY_APP_DEV_MODE_HELM_CHART=/path/to/blah ./myapp.py
```

File: README.md [hunk 2]
```diff
@@ -139,6 +139,96 @@ command groups. The `NullOrchestrator` disables commands related to managing ser
...
+
+1. Globally for any files dumped in the `values` directory.
+2. For a set key by dumping files in `values-files` at that nested key layer.
+
+Take the following `cli/helm/` directory structure:
+
+```bash
+drwxrwxr-x  cli/helm/
+drwxrwxr-x ├──  values/
+.rw-rw-r-- │  └──  foobar.yml
```
Comment (al-jeyapal) [inline README.md:206]: Include another file in here, so it's obvious that they can have multiple files.

File: README.md [hunk 3]
```diff
@@ -139,6 +139,96 @@ command groups. The `NullOrchestrator` disables commands related to managing ser
...
+
+##### Kubeconfig
+
+A custom `kubeconfig` file can be used by specifying the `KUBECONFIG` environment variable.
+
+```bash
+KUBECONFIG=/opt/brightsparklabs/myapp/conf/.generated/config ./myapp ...
+```
+
+_NOTE:_ The `KUBECONFIG` file must be one that is visible to the container.
```
Comment (al-jeyapal) [inline README.md:230]: Need to expand on this.

... Which means it must be at a location which is mounted into the launch container. Refer to section <something> for details on what volumes are mounted into the launch container.

File: appcli/dev_mode.py [hunk 4]
```diff
@@ -30,7 +30,7 @@ def wrap_dev_mode():
       The below code:
 
         if dev_mode_enabled:
-          with wrap_dev_mode:
+          with wrap_dev_mode():
             ...
             logger.info("DEV MODE operation successfuly")
```
Comment (al-jeyapal) [inline appcli/dev_mode.py:35]: There's a typo. Let's just add a better message:

```
            logger.info("DEV MODE operation completed successfully")		￼

File: appcli/models/cli_context.py [hunk 5]
```diff
@@ -31,6 +31,24 @@
...
+"""
+
+HELM_VALUES_FILES = HELM_DIRECTORY / "values-files"
+""" The directory containing all key-specific files.
+All parent directories for a file are used as nested keys.
+The [0] filename component (split on `.`) is used for the key.
+
+All files in this directory are applied with:
+    --set-file path.to.foo=/path/to/foo.bar.yml
+"""
```
Comment (al-jeyapal) [inline appcli/models/cli_context.py:50]: This is hard to follow. It's worth putting in the actual example similar to the README which is really clear.

File: appcli/dev_mode.py (hunk 4 above)
Comment (al-jeyapal) [inline appcli/dev_mode.py:35]: typo

```
logger.info("DEV MODE operation completed successfully")
```

File: README.md [hunk 6]
```diff
@@ -139,6 +139,96 @@ command groups. The `NullOrchestrator` disables commands related to managing ser
...
+    # Name of the project.
+    release_name="myapp",
+
+    # Chart archive path (relative to `conf/templates/cli/helm/`).
+    # [Optional] Default is `chart/`
+    chart_location="mychart.tgz",
+
+    # Chart source path (absolute).
+    # [Optional] Default is `None`.
+    dev_chart_location="path/to/mychart"
```
Comment (al-jeyapal) [inline README.md:180]: I don't really see the value of including this in the config. It's only for DEV purposes, we shouldn't ever publish an app with a hard-coded path from a dev's workstation, so people will always need to remember to clean up the file.

Let's just use an env var instead:
```
MYAPP_DEV_MODE=true MYAPP_DEV_MODE_HELM_CHART_PATH=/path/to/mychart ./myapp.py
```

We also need to explain it better. I.e. precisly why you added it. During development it would be slow to require packaging up the chart for any change, etc. This mechanism allows developers to iterate on the chart without needing to blah blah blah

File: README.md (hunk 3 above)
Comment (al-jeyapal) [inline README.md:230]: Need to explain this better.
```
... I.e. In a location which gets mounted in the launch container. Refer to <section ..> for an explanation of what volumes are mounted into the launch container.
```

Hopefully we already have <section> somewhere in the README which explains it. Otherwise add it.

File: appcli/models/cli_context.py [hunk 7]
```diff
@@ -217,3 +235,34 @@ def get_variables_manager(self) -> VariablesManager:
...
+        """
+        return self.get_generated_configuration_dir().joinpath(HELM_VALUES)
+
+    def get_helm_values_files_dir(self) -> Path:
+        """The directory containing all key-specific files.
+        All parent directories for a file are used as nested keys.
+        The [0] filename component (split on `.`) is used for the key.
+
+        All files in this directory are applied with:
+            --set-file path.to.foo=/path/to/foo.bar.yml
```
Comment (al-jeyapal) [inline appcli/models/cli_context.py:263]: Explain it better as above.

File: appcli/models/cli_context.py [hunk 8]
```diff
@@ -217,3 +235,34 @@ def get_variables_manager(self) -> VariablesManager:
             key_file=self.get_key_file(),
             application_context_files_dir=self.application_context_files_dir,
         )
+
+    def get_helm_dir(self) -> Path:
+        """Parent directory for all helm related configs.
```
Comment (al-jeyapal) [inline appcli/models/cli_context.py:240]: I don't love the fact that these things are in the root CLI context. They make no sense when not using the Helm orchestrator.

We should move the three of them into the `HelmOrchestrator` class itself. Just like how the `DockerComposeOrchestrator` handles its own specific paths (`docker_compose_file`, `docker_compose_override_directory`, etc).

File: README.md [hunk 9]
```diff
@@ -139,6 +139,96 @@ command groups. The `NullOrchestrator` disables commands related to managing ser
...
+.rw-------                └──  config  # Overwrite this with a cluster specific config file. ie `~/.kube/config`.
+```
+
+You can then define the orchestrator:
+
+```python
+from appcli.orchestrators import HelmOrchestrator
+orchestrator = HelmOrchestrator(
+    # Name of the project.
+    release_name="myapp",
```
Comment (al-jeyapal) [inline README.md:172]: Why do we want to set another name here? Can't we just use the app name from the `Configuration` class?

Or even `CliContext.get_project_name` if its appropriate to include the release environment to distinguish things.

File: appcli/orchestrators.py [hunk 10]
```diff
@@ -598,6 +599,239 @@ def __exec_command(self, command: Iterable[str]) -> CompletedProcess:
...
+        Returns:
+            CompletedProcess: Result of the orchestrator command.
+        """
+        # Generate the command string.
+        command = [
+            "helm",
+            "upgrade",
+            "--install",
+            "--namespace",
+            self.release_name,
```
Comment (al-jeyapal) [inline appcli/orchestrators.py:646]: Yeh, def think it should be pulled from `cli_context.configuration....` or `cli_context.get_project_name()`

File: appcli/orchestrators.py [hunk 11]
```diff
@@ -598,6 +599,239 @@ def __exec_command(self, command: Iterable[str]) -> CompletedProcess:
...
+            "--create-namespace",
+        ]
+        # Set values args.
+        for arg in self.__generate_values_args(
+            helm_values_dir=cli_context.get_helm_values_dir(),
+            helm_values_files_dir=cli_context.get_helm_values_files_dir(),
```
Comment (al-jeyapal) [inline appcli/orchestrators.py:652]: Another benefit of this being instance vars in the class, is that we can calculate the list of args in the contructor.

File: appcli/orchestrators.py [hunk 12]
```diff
@@ -598,6 +599,239 @@ def __exec_command(self, command: Iterable[str]) -> CompletedProcess:
...
+        ):
+            command.append(arg)
+        # Set release name.
+        command.append(self.release_name)
+        # Set chart location.
+        # If we're in `dev_mode` and `dev_chart_location` is set, use that.
+        if cli_context.is_dev_mode and self.dev_chart_location is not None:
+            with wrap_dev_mode():
+                logger.debug(f"Deploying chart from `{self.dev_chart_location}`")
```
Comment (al-jeyapal) [inline appcli/orchestrators.py:661]: Maybe for clarity :
```
logger.debug(f"Found DEV_MODE chart. Ignoring bundled chart from `{self.helm_dir / self.chart_location}"`)
logger.debug(f"Deploying chart from `{self.dev_chart_location}`)
```

File: README.md [hunk 13]
```diff
@@ -139,6 +139,96 @@ command groups. The `NullOrchestrator` disables commands related to managing ser
...
+    # [Optional] Default is `None`.
+    dev_chart_location="path/to/mychart"
+)
+```
+
+_NOTE:_ Using `dev_chart_location` allows you to specify the location to your charts source code, so that live changes can be reflected every time `./myapp service start` is run.
+The following conditions must be met in order to use it:
+
+1. `dev_chart_location` is set, and is an **ABSOLUTE** path to the chart directory.
+2. Your application must be running in `dev_mode` as validated by:
```
Comment (al-jeyapal) [inline README.md:188]: Use `DEV_MODE` when referring to it since that's what we prefix lines with (i.e. in capitals)

File: appcli/orchestrators.py [hunk 14]
```diff
@@ -598,6 +599,239 @@ def __exec_command(self, command: Iterable[str]) -> CompletedProcess:
...
+        """
+        Creates a new instance of an orchestrator for helm-based applications.
+
+        NOTE: All `Path` objects are relative to the configuration directory.
+
+        Args:
+            release_name (str): A name to give to this helm release.
+            chart_location (Path): Path to the helm chart file/directory to deploy.
+            dev_chart_location (Path | None):
+                [Optional] An absolute path to the chart source for when deploying in `dev_mode`.
```
Comment (al-jeyapal) [inline appcli/orchestrators.py:622]: ... when deploying in `DEV_MODE`

File: appcli/orchestrators.py [hunk 15]
```diff
@@ -598,6 +599,239 @@ def __exec_command(self, command: Iterable[str]) -> CompletedProcess:
...
+        # Set values args.
+        for arg in self.__generate_values_args(
+            helm_values_dir=cli_context.get_helm_values_dir(),
+            helm_values_files_dir=cli_context.get_helm_values_files_dir(),
+        ):
+            command.append(arg)
+        # Set release name.
+        command.append(self.release_name)
+        # Set chart location.
+        # If we're in `dev_mode` and `dev_chart_location` is set, use that.
```
Comment (al-jeyapal) [inline appcli/orchestrators.py:658]: If we're in `DEV_MODE` ...

Comment (al-jeyapal) [reply appcli/orchestrators.py:658]: As above, let's use an env var `<APPNAME>_DEV_MODE_HELM_CHART_PATH` instead.

File: appcli/orchestrators.py [hunk 16]
```diff
@@ -598,6 +599,239 @@ def __exec_command(self, command: Iterable[str]) -> CompletedProcess:
...
+        self, cli_context: CliContext, service_names: tuple[str, ...]
+    ) -> bool:
+        if service_names is None or len(service_names) == 0:
+            return True
+        logger.info("HelmOrchestrator has no services.")
+        return False
+
+    def get_disabled_commands(self) -> list[str]:
+        # The `task` command is disabled because helm is only used for install/upgrade/uninstall/downgrade operations.
+        # The `init` command is disabled because it's used to initialise additional services, and this orchestrator
```
Comment (al-jeyapal) [inline appcli/orchestrators.py:765]: Reorder the comments to align with the orders of commands in the array below.

File: appcli/orchestrators.py [hunk 17]
```diff
@@ -598,6 +599,239 @@ def __exec_command(self, command: Iterable[str]) -> CompletedProcess:
...
+        """Run the given command and return the CompletedProcess.
+
+        Args:
+            command (list[str]): The command to execute.
+
+        Returns:
+            CompletedProcess: The execution result.
+        """
+        logger.debug(f"Executing {str(command)}")
+        result = subprocess.run(command, capture_output=True)
```
Comment (al-jeyapal) [inline appcli/orchestrators.py:779]: We don't want to use `capture_output=False` so that it live prints progress? Isn't that what we do in the docker compose one?
Otherwise, isn't it just going to block for a long time while we have no idea what's going on?

File: appcli/orchestrators.py [hunk 18]
```diff
@@ -598,6 +599,239 @@ def __exec_command(self, command: Iterable[str]) -> CompletedProcess:
...
+            raise subprocess.CalledProcessError(
+                result.returncode,
+                command,
+                result.stdout,
+                result.stderr,
+            )
+        return result
+
+    def __generate_values_args(
+        self, helm_values_dir: Path, helm_values_files_dir: Path
```
Comment (al-jeyapal) [inline appcli/orchestrators.py:791]: You won't need to pass these in anymore as they will be part of `self.helm_values_dir`, etc

File: appcli/orchestrators.py [hunk 19]
```diff
@@ -598,6 +599,239 @@ def __exec_command(self, command: Iterable[str]) -> CompletedProcess:
...
+    ) -> list[str | Path]:
+        """Recursively takes all the values files in the directories and generates an args array to
+        pass to helm through either `--values` or `--set-file` (depending on the location).
+        e.g:
+
+            ["--values", "values.yaml", "--set-file", "path.to.foo=path/to/foo.yaml"...
+
+        Args:
+            helm_values_dir (Path): Directory containing files to set with `--values`.
+            helm_values_files_dir (Path): Directory containing files to set with `--set-file`.
```
Comment (al-jeyapal) [inline appcli/orchestrators.py:801]: Add you example from the readme for clarity.

File: appcli/orchestrators.py [hunk 20]
```diff
@@ -598,6 +599,239 @@ def __exec_command(self, command: Iterable[str]) -> CompletedProcess:
...
+
+        Returns:
+            list[str | Path]: The arg list.
```
Comment (al-jeyapal) [inline appcli/orchestrators.py:804]: Better to just be `list[str]`. We don't really want to return `Path` in here do we? Its weird having a mix of types.
We're using this to build up the command string yeah? So we should probably be turning the Path objects into strings.

File: appcli/templates/launcher.j2 [hunk 21]
```diff
@@ -63,7 +64,7 @@ function main()
         --volume "{{ path }}:{{ path }}" \
 {% endfor %}
         --volume "${DOCKER_SOCKET}:/var/run/docker.sock" \
-        --volume "${MOUNTED_DATA_DIR}/cli/home:/root" \
+        --volume "${MOUNTED_GENERATED_CONFIG_DIR}/cli/home:/root" \
```
Comment (al-jeyapal) [inline appcli/templates/launcher.j2:67]: This isn't great.
A home directory could contain actual data depending on what the system does.
What if something in the app writes data into the home dir that we need in future? We will lose it if the config dir gets regenerated.

Can we instead copy the required helm stuff from the generated directory into the home directory once container is launched?

**Review by al-jeyapal — CHANGES_REQUESTED** [review]

File: appcli/orchestrators.py (hunk 11 above)
Comment (benjamin-wiffen) [reply appcli/orchestrators.py:652]: We can't do it in the constructor as it doesn't get access to `cli_context.get_generated_configuration_dir()` which we need to find the absolute path to the files.

File: appcli/orchestrators.py [hunk 22]
```diff
@@ -598,6 +599,271 @@ def __exec_command(self, command: Iterable[str]) -> CompletedProcess:
...
+        Args:
+            command (list[str]): The command to execute.
+
+        Returns:
+            CompletedProcess: The execution result.
+        """
+        logger.debug(f"Executing {str(command)}")
+        result = subprocess.run(command, capture_output=False)
+        if result.returncode != 0:
+            if result.stderr is not None:
```
Comment (al-jeyapal) [inline appcli/orchestrators.py]: Since we're now doing `capture_output=False`, so `result.stdout` and `result.stderr` are not captured. I'll get tid of this stuff.

File: appcli/orchestrators.py (hunk 20 above)
Comment (al-jeyapal) [reply appcli/orchestrators.py]: Looks like you've changed it to be `str(Path)` but not updated method signature. Will fix.

**Review by al-jeyapal — APPROVED** [review]
